test/powershell/engine/Api/TaskBasedAsyncPowerShellAPI.Tests.ps1

```diff
@@ -0,0 +1,251 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+Describe 'Task-based PowerShell async APIs' -Tags 'CI' {
+    BeforeAll {
+        $sbStub = @'
+.foreach{
+    [pscustomobject]@{
+        Time       = [DateTime]::Now.ToString('yyyyMMddTHHmmss.fffffff')
+        Value      = $_
+        ThreadId   = [System.Threading.Thread]::CurrentThread.ManagedThreadId
+        RunspaceId = [runspace]::DefaultRunspace.Id
+    }
+    Start-Sleep -Milliseconds 500
+}
+'@
+    }
+
+    Context 'PowerShell::InvokeAsync - Single script tests' {
+        BeforeAll {
+            function InvokeAsyncHelper {
+                param(
+                    [powershell]$PowerShell,
+                    [switch]$Wait
+                )
+                $r = $PowerShell.AddScript("@(1,2,3,4,5)${sbStub}").InvokeAsync()
+                if ($Wait) {
+                    [System.Threading.Tasks.Task]::WaitAll(@($r))
+                }
+                $r
+            }
+
+            function GetInnerErrorId {
+                param(
+                    [exception]$Exception
+                )
+                while ($null -ne $Exception.InnerException) {
+                    $Exception = $Exception.InnerException
+                    if ($null -ne (Get-Member -InputObject $Exception -Name ErrorRecord)) {
+                        $Exception.ErrorRecord.FullyQualifiedErrorId
+                        break
+                    }
+                }
+            }
+        }
+
+        It 'can invoke a single script asynchronously, but only in a new runspace' {
+            $ps = [powershell]::Create()
+            try {
+                $r = InvokeAsyncHelper -PowerShell $ps -Wait
+                $r.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+                $r.IsCompletedSuccessfully | Should -Be $true
+            } finally {
+                $ps.Dispose()
+            }
+        }
+
+        It 'handles terminating errors properly when invoked asynchronously in a new runspace' {
+            $ps = [powershell]::Create()
+            try {
+                $sb = {
+                    $r = $ps.AddScript(@'
+try {
+    Get-Process -Invalid 42 -ErrorAction Stop
```

Comment:
`-ErrorAction stop` is not needed here because the exception is thrown from the parameter binder, which is terminating.

Reply:
That's just not true. Compare:

```
[powershell]::Create().AddScript('gps -invalid 42 -ea stop').Invoke()
```

with

```
[powershell]::Create().AddScript('try {gps -invalid 42 -ea stop} catch {throw}').Invoke()
```

You need try/catch to make the terminating error actually terminate the script. Otherwise there is no exception, no faulted task to check for.

Reply:
I was not talking about the try/catch/throw part of it. The comment was on `-ErrorAction Stop`

Reply:
Gotcha, sorry I misunderstood. Corrected.